collector/lib/ProcfsScraper.cpp

```diff
@@ -367,7 +412,37 @@ bool GetConnections(int dirfd, UnorderedMap<ino_t, ConnInfo>* connections, Unord
     FDHandle net_tcp6_fd = openat(dirfd, "net/tcp6", O_RDONLY);
     if (net_tcp6_fd.valid()) {
       FileHandle net_tcp6(std::move(net_tcp6_fd), "r");
-      success = ReadConnectionsFromFile(Address::Family::IPV6, L4Proto::TCP, net_tcp6, connections, listen_endpoints) && success;
+      success = ReadConnectionsAndEndpointsFromFile(Address::Family::IPV6, net_tcp6, L4Proto::TCP, connections, listen_endpoints) && success;
+    } else {
+      success = false;
+    }
+  }
+
+  {
+    FDHandle net_udp_fd = openat(dirfd, "net/udp", O_RDONLY);
+    if (net_udp_fd.valid()) {
+      FileHandle net_udp(std::move(net_udp_fd), "r");
+      success = ReadConnectionsAndEndpointsFromFile(Address::Family::IPV4, net_udp, L4Proto::UDP, connections, listen_endpoints) && success;
+    } else {
+      success = false;  // there should always be a net/udp file
+    }
+  }
+
+  {
+    FDHandle net_udp6_fd = openat(dirfd, "net/udp6", O_RDONLY);
+    if (net_udp6_fd.valid()) {
+      FileHandle net_udp6(std::move(net_udp6_fd), "r");
+      success = ReadConnectionsAndEndpointsFromFile(Address::Family::IPV6, net_udp6, L4Proto::UDP, connections, listen_endpoints) && success;
+    } else {
+      success = false;
+    }
+  }
```

Comment:
Seems like this can be refactored. We could loop through the files instead of repeating all of this code. I would do the refactor, but I am not sure about the correctness of the code as is. I am not sure that it is correct to set success to false in all of these cases, if valid() is false. That was done for both net/tcp and net/tcp6. Should it only have been done for net/tcp?

Reply:
I think you're right - a loop might be a fair bit neater (and I think we can use a list of `std::tuple` to store the various bits of data that each invocation needs (filename, family, protocol))

On the point about `fd.valid()`, I think that condition holds for all the tcp and udp files since they should always exist (unless the kernel is configured without procfs, or without the necessary networking stacks but that feels very unlikely for our target kernels) There may still be some benefit to adding more granular error reporting between the two protocols, but that will mean splitting this function in two (tcp & udp) and letting the caller call both, and decide how to handle the errors. Not necessarily something we need to do now.

Reply:
I ended up looping through the files and using a tuple though the tuple just holds filename and protocol.